lib/route/seg6.c

```diff
@@ -0,0 +1,76 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+
+#include <stdbool.h>
+#include <linux/seg6.h>
+#include <linux/ipv6.h>
+#include <netlink/utils.h>
+#include <netlink-private/utils.h>
+
+bool seg6_validate_srh(struct ipv6_sr_hdr *srh, int len, bool reduced)
+{
+	unsigned int tlv_offset;
+	int max_last_entry;
+	int trailing;
+
+	if (srh->type != IPV6_SRCRT_TYPE_4)
+		return false;
+
+	if (((srh->hdrlen + 1) << 3) != len)
+		return false;
+
+	if (!reduced && srh->segments_left > srh->first_segment) {
+		return false;
+	} else {
+		max_last_entry = (srh->hdrlen / 2) - 1;
+
+		if (srh->first_segment > max_last_entry)
+			return false;
+
+		if (srh->segments_left > srh->first_segment + 1)
+			return false;
+	}
+
+	tlv_offset = sizeof(*srh) + ((srh->first_segment + 1) << 4);
+
+	trailing = len - tlv_offset;
+	if (trailing < 0)
+		return false;
+
+	while (trailing) {
+		struct sr6_tlv *tlv;
+		unsigned int tlv_len;
+
+		if (trailing < sizeof(*tlv))
+			return false;
+
+		tlv = (struct sr6_tlv *)((unsigned char *)srh + tlv_offset);
+		tlv_len = sizeof(*tlv) + tlv->len;
+
+		trailing -= tlv_len;
+		if (trailing < 0)
+			return false;
+
+		tlv_offset += tlv_len;
+	}
+
+	return true;
+}
+
+void seg6_dump_srh(struct nl_dump_params *dp, struct ipv6_sr_hdr *srh)
+{
+	int i, offset;
+	char addr[INET6_ADDRSTRLEN];
+	struct sr6_tlv_hmac *tlv;
+
+	nl_dump(dp, "segs %d [ ", srh->first_segment + 1);
+	for (i = srh->first_segment; i >= 0; i--) {
+		nl_dump(dp, "%s", _nl_inet_ntop(AF_INET6, &srh->segments[i], addr));
+	}
+	nl_dump(dp, "] ");
+	if (sr_has_hmac(srh)) {
+		offset = ((srh->hdrlen + 1) << 3) - 40;
```

Comment:
where does the 40 come from? The entire expression is rather hard to understand. It should be simplified, possibly using a define for non-obvious things. Or at least a code comment, that explains where thsoe numbers come from.

Reply:
40 changed to sizeof(..)